### PR TITLE
POST for Listings

### PR DESCRIPTION
**Pull request by rekibnikufesin**

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -1,10 +1,101 @@
-web3==5.0.2
```
Comment (rekibnikufesin) [inline requirements.txt]: I did a `pip freeze > requirements.txt` here. It added all subdependencies.
Alternatively, we could just add the packages I added (celery, redis) to this file and reduce the number of packages tracked in it

File: apis/listing/listings.py [hunk 2]
```diff
@@ -89,29 +105,32 @@ def get_filename(self):
...
-
-        if not os.path.exists(dest):
-            os.makedirs(dest)
-
-        loc = f'{dest}{listing_hash}'
-
-        item.save(loc)
 
-        with open(loc, 'rb') as data:
+        # Read the file the first time to verify the md5 of the uploaded file
```
Comment (rekibnikufesin) [inline apis/listing/listings.py:128]: Not thrilled here about reading the same file 3 times, but don't have a clean alternative.
First read: verify the md5sum with what the client provided to verify we received the complete file
Second read: upload to S3
Third read: calculate the keccak so we can set the datahash

File: core/celery.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+@celery.task
+def send_hash_after_mining(listing, data_hash):
+    """
+    Wait for the transaction to be mined, then set the data_hash
+    for the uploaded file in protocol
+    """
+    #TODO: calculate hash for listing
+    #TODO: wait for transaction to be mined
+    #TODO: set the data hash
+    if is_registered == True:
```
Comment (robrobbins) [inline core/celery.py]: `is_registered` is a method so -> `is_registered()`

File: core/celery.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        dt = get_datatrust()
+        t = dt.set_data_hash(listing, data_hash)
+        # TEST env never need use send...
+        if current_app.config['TESTING'] == True:
+            tx = transact(t)
+        else:
+            tx = send(g.w3, current_app.config['PRIVATE_KEY'], t)
```
Comment (robrobbins) [inline core/celery.py]: i'll prob end up abstracting this into a protocol helper `send_or_transact` or some shit. just talkin to myself here...

File: requirements.txt (hunk 1 above)
Comment (robrobbins) [reply requirements.txt]: oh yeah, fuck this . just add celery and redis by hand. pip freeze is what got me into trouble with the last one when i went to change something...

File: tests/listing/test_listings.py [hunk 5]
```diff
@@ -126,3 +127,24 @@ def test_get_listings(w3, market_token, voting, parameterizer_opts, datatrust, l
     assert payload['items'][0]['listing_hash'] == w3.toHex(listing_hash) # payload hashes are hex
     assert payload['items'][0]['title'] == 'lol catz 9000'
     assert payload['to_block'] > 0
+
+def test_post_listings(w3, voting, datatrust, listing, test_client, s3_client):
+    listing = test_client.post('/listings/', 
+    content_type='multipart/form-data',
```
Comment (robrobbins) [inline tests/listing/test_listings.py:141]: nice. i was wondering how to mock the form. :heavy_check_mark:

File: tests/listing/test_listings.py [hunk 6]
```diff
@@ -126,3 +127,24 @@ def test_get_listings(w3, market_token, voting, parameterizer_opts, datatrust, l
...
+        md5_sum='7f7c47e44b125f2944cb0879cbe428ce',
+        listing_hash='1234',
+        file=(BytesIO(b'a pony'), 'my_little_pony.gif')
+    ))
+    assert listing.status_code == 201
+    
+    uploaded_file = g.s3.get_object(
+        Bucket=current_app.config['S3_DESTINATION'],
+        Key='1234'
+    )['Body'].read().decode()
```
Comment (robrobbins) [inline tests/listing/test_listings.py:161]: fully mocked s3, celery, and blockchain when its done...

_be still my heart_

File: apis/listing/listings.py (hunk 2 above)
Comment (robrobbins) [reply apis/listing/listings.py:128]: go ahead an flush it out with the *way too many reads* first. that way we know it works.

once everything is in i'll stare it for a cpl hours, be easier to see then

**Pull request merged**